### PR TITLE
Gitignore artifacts. Add DepositContract and modify the DepositFactoryContract in order to create multi instances of DepositContract by the clone pattern

### PR DESCRIPTION
**Pull request by NguyetPhamFuix**
closes [#109](https://github.com/Sprynt-io/Sprynt-Mint-Front-End/issues/109), [#141](https://github.com/Sprynt-io/Sprynt-Mint-Front-End/issues/141), [#142](https://github.com/Sprynt-io/Sprynt-Mint-Front-End/issues/142), [#143](https://github.com/Sprynt-io/Sprynt-Mint-Front-End/issues/143)

### CODE REVIEW

**Comment by ngmitam** [comment]
Please write unit tests whenever you can, thanks

**Comment by NguyetPhamFuix** [comment]
> Please write unit tests whenever you can, thanks

I have updated the unit tests to make sure the defined flow run well. Thank you very much.

File: contracts/childs/DepositContract.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+import {DepositItem} from "../library/Structs.sol";
+import {DepositFactoryContract} from "../parents/DepositFactoryContract.sol";
+
+contract DepositContract {
+    address private _sellerAddress;
+    address public tokenAddress;
```
Comment (ngmitam) [inline contracts/childs/DepositContract.sol:8]: should have minted to keep track number of minted token

File: contracts/childs/DepositContract.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        _factoryContractAddress = factoryContractAddress;
+    }
+
+    function mint(
+        DepositItem calldata depositItem,
+        bytes calldata signature,
+        uint256 lzGasFee,
+        bool isNativeToken
+    ) public payable {
+        DepositFactoryContract(_factoryContractAddress).depositTokenByClient{
```
Comment (ngmitam) [inline contracts/childs/DepositContract.sol:72]: Please check totalSupply, mintQuantity, dstChainId, selleraddress, tokenAddress, ...

File: contracts/childs/DepositContract.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    modifier onlyPermissioned() {
+        require(
+            msg.sender == _sellerAddress ||
+                msg.sender == _factoryContractAddress,
+            "No permission!"
+        );
+        _;
+    }
+
+    function changeMintPrice(uint256 mintPrice_) public onlyPermissioned {
```
Comment (ngmitam) [inline contracts/childs/DepositContract.sol:86]: For each function can call by factoryContract please create a function for factoryContract, these function sould only call by configure role

**Review by ngmitam — CHANGES_REQUESTED** [review]

File: contracts/childs/DepositContract.sol (hunk 2 above)
Comment (NguyetPhamFuix) [reply contracts/childs/DepositContract.sol:72]: These information are included in the DepositItem struct.

File: contracts/childs/DepositContract.sol (hunk 3 above)
Comment (NguyetPhamFuix) [reply contracts/childs/DepositContract.sol:86]: Could you explain this more. I have create the "onlyPermissioned" modifier to set that these functions can only be called by seller and factoryContract. So how can I modify this as your requirement.
image

Comment (ngmitam) [reply contracts/childs/DepositContract.sol:86]: These function can be called by factoryContract, but now factoryContract don't have functions to call them. So please create functions on factoryContract

File: contracts/childs/DepositContract.sol (hunk 2 above)
Comment (ngmitam) [reply contracts/childs/DepositContract.sol:72]: Yes, but check these info with contract's config.

File: contracts/childs/DepositContract.sol (hunk 3 above)
Comment (NguyetPhamFuix) [reply contracts/childs/DepositContract.sol:92]: Ah I see. Thank you. I will fix it now

**Comment by ngmitam** [comment]
Should also keep track the minted token on receiver contract and generate token id automatically, dont allow user to set tokenId. We have multi deposit chains.

**Comment by ngmitam** [comment]
We use ERC1155 for NFT, so 1 tokenId only supply 1.

File: contracts/parents/DepositFactoryContract.sol [hunk 4]
```diff
@@ -9,139 +9,343 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
-  
-  function estimateFee(uint16 dstChainId_, bool _useZro, bytes calldata _adapterParams) public view returns (uint nativeFee, uint zroFee) {
-    bytes memory encodedPayload = abi.encode(
-      address(0),
-      uint256(0),
-      address(0)
-    );
-    return lzEndpoint.estimateFees(dstChainId_, address(this), encodedPayload, _useZro, _adapterParams);
-  }
+    function setupDepositRole(
```
Comment (ngmitam) [inline contracts/parents/DepositFactoryContract.sol:145]: Should also have revokeRole

File: contracts/childs/ERC1155.sol [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract ERC1155Contract is ERC1155, AccessControl, ReentrancyGuard {
+    bytes32 public constant FACTORY_CONTRACT_ROLE =
+        keccak256("FACTORY_CONTRACT_ROLE");
+
+    bytes32 public constant OWNER_ROLE = keccak256("OWNER_ROLE");
+    string private baseURI;
+
+    mapping(address => uint256) private _mintedTokens;
```
Comment (ngmitam) [inline contracts/childs/ERC1155.sol:15]: It is just a number, no need to map with address

File: contracts/parents/ReceiveFactoryContract.sol [hunk 6]
```diff
@@ -1,65 +1,84 @@
...
 import "@layerzerolabs/solidity-examples/contracts/lzApp/NonblockingLzApp.sol";
-import "@layerzerolabs/solidity-examples/contracts/util/BytesLib.sol";
-import "../childs/PolarysNftContract.sol";
-import "../interfaces/IPolarysNftContract.sol";
-import "hardhat/console.sol";
+import "../childs/ERC1155.sol";
 
 contract ReceiveFactoryContract is NonblockingLzApp {
-  using BytesLib for bytes;
+    uint256 private _mintedTokens;
```
Comment (ngmitam) [inline contracts/parents/ReceiveFactoryContract.sol:7]: Keep track of the minted on the NFT contract, not here

File: contracts/parents/ReceiveFactoryContract.sol [hunk 7]
```diff
@@ -1,65 +1,84 @@
...
+        (
+            address clientAddress,
+            uint256 mintQuantity,
+            bytes memory data,
+            address seller
+        ) = abi.decode(_payload, (address, uint256, bytes, address));
+        address nftContractAddress = nftContracts[seller];
+        require(nftContractAddress != address(0), "NftContract is not created");
+
+        uint256[] memory ids = new uint256[](mintQuantity);
```
Comment (ngmitam) [inline contracts/parents/ReceiveFactoryContract.sol:56]: Override the minBatchToken of NFT contract, don't implement logic here

File: contracts/parents/DepositFactoryContract.sol [hunk 8]
```diff
@@ -9,139 +9,362 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
+            address(0x0),
+            adapterParams,
+            lzGasFee
+        );
+
+        if (isNativeToken) {
+            Address.sendValue(payable(_adminWallet), platformFeeAmount);
+            Address.sendValue(payable(depositItem.sellerAddress), userProfit);
+        } else {
+            IERC20(_acceptToken).safeTransferFrom(
```
Comment (ngmitam) [inline contracts/parents/DepositFactoryContract.sol:262]: Should allow seller to set acceptToken on their depositContract and use it

File: contracts/parents/DepositFactoryContract.sol [hunk 9]
```diff
@@ -9,139 +9,362 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
+    }
+
+    function _nonblockingLzReceive(
+        uint16,
+        bytes memory,
+        uint64,
+        bytes memory _payload
+    ) internal override {
+        (
+            address sellerAddress,
```
[the remark this answers is not in the document]
Comment (ngmitam) [reply contracts/parents/DepositFactoryContract.sol:89]: It should be a map
mapping(address => address) public sellerToDepositContract;

File: contracts/childs/DepositContract.sol [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        _factoryContractAddress = factoryContractAddress;
+    }
+
+    function mint(
+        DepositItem calldata depositItem,
+        bytes calldata signature,
+        uint256 lzGasFee,
+        bool isNativeToken,
+        bytes calldata adapterParams
+    ) public payable {
```
Comment (ngmitam) [inline contracts/childs/DepositContract.sol:74]: check mintPrice
if whiteList use whiteListMintPrice

**Review by ngmitam — APPROVED** [review]

File: contracts/parents/DepositFactoryContract.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (ngmitam) [reply contracts/parents/DepositFactoryContract.sol:89]: @NguyetPhamFuix

File: contracts/parents/DepositFactoryContract.sol (hunk 8 above)
Comment (ngmitam) [reply contracts/parents/DepositFactoryContract.sol:262]: @NguyetPhamFuix

File: contracts/childs/DepositContract.sol (hunk 10 above)
Comment (ngmitam) [reply contracts/childs/DepositContract.sol:74]: @NguyetPhamFuix

File: contracts/parents/ReceiveFactoryContract.sol [hunk 11]
```diff
@@ -1,65 +1,178 @@
...
+            dstChainId,
+            encodedPayload,
+            payable(tx.origin),
+            address(0x0),
+            adapterParams,
+            nativeFee
+        );
+    }
+
+    function _nonblockingLzReceive(
```
Comment (ngmitam) [inline contracts/parents/ReceiveFactoryContract.sol:128]: We will no longer use LayerZero to mint, instead, we call directly by our account.

File: contracts/childs/DepositContract.sol [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        deadline = deadline_;
+        _factoryContractAddress = factoryContractAddress;
+        if (whiteList_.length > 0) {
+            for (uint256 i = 0; i < whiteList_.length; i++) {
+                whiteList[whiteList_[i]] = true;
+            }
+        }
+    }
+
+    function mint(
```
Comment (ngmitam) [inline contracts/childs/DepositContract.sol:82]: We will no longer use LayerZero

File: contracts/parents/DepositFactoryContract.sol [hunk 13]
```diff
@@ -9,139 +9,574 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
+
+    modifier onlyContract() {
+        require(
+            _depositContracts[msg.sender],
+            "Only contracts created by this factory can call the function!"
+        );
+        _;
+    }
+
+    function depositTokenByClient(
```
Comment (ngmitam) [inline contracts/parents/DepositFactoryContract.sol:393]: We will no longer user LayerZero

File: contracts/parents/DepositFactoryContract.sol [hunk 14]
```diff
@@ -9,139 +9,574 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
+            payContracts[msg.sender] == address(0),
+            "already created pay contract."
+        );
 
-  function _verify(bytes32 digest, bytes memory signature)
-  internal view returns (bool)
-  {
-    return hasRole(DEPOSIT_ROLE, ECDSA.recover(digest, signature));
-  }
+        SimplePay simplePayContract = new SimplePay(
```
Comment (ngmitam) [inline contracts/parents/DepositFactoryContract.sol:212]: Use clone factory

File: contracts/parents/DepositFactoryContract.sol [hunk 15]
```diff
@@ -9,139 +9,574 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
+    //                 deadline,
+    //                 address(this),
+    //                 whiteList_
+    //             );
+    //             _depositContracts[clone] = true;
+    //             _depositContractsList.push(clone);
+    //             deployedDepositContracts[dstChainId][sellerAddress] = clone;
+    //             emit DepositContractCreated(clone);
+    //         } else {
+    //             address masterDepositContractAddress = address(
```
Comment (ngmitam) [inline contracts/parents/DepositFactoryContract.sol:318]: always use clone factory, don't use new

File: contracts/parents/ReceiveFactoryContract.sol [hunk 16]
```diff
@@ -1,65 +1,178 @@
...
+    constructor(
+        address _layerZeroEndpoint
+    ) NonblockingLzApp(_layerZeroEndpoint) {}
+
+    function createNftContractBySeller(string calldata tokenUri) external {
+        require(
+            nftContracts[msg.sender] == address(0),
+            "already created nft contract."
+        );
+        ERC1155Contract newNftContract = new ERC1155Contract(
```
Comment (ngmitam) [inline contracts/parents/ReceiveFactoryContract.sol:43]: Alway use clone factory

File: contracts/childs/ERC1155.sol [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+            "contract address must not be zero address"
+        );
+        _grantRole(FACTORY_CONTRACT_ROLE, factoryContractAddress);
+    }
+
+    function mintToken(
+        address to,
+        uint256 id,
+        bytes memory data
+    ) external onlyRole(FACTORY_CONTRACT_ROLE) nonReentrant {
```
Comment (ngmitam) [inline contracts/childs/ERC1155.sol:54]: Should allow our account to mint Token (MINT ROLE or ...)

File: contracts/childs/ERC1155.sol [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        require(to != address(0), "Address must not be zero address");
+        _mint(to, id, 1, data);
+    }
+
+    function mintBatchToken(
+        address to,
+        uint256 quantity,
+        bytes memory data
+    ) external onlyRole(FACTORY_CONTRACT_ROLE) nonReentrant {
```
Comment (ngmitam) [inline contracts/childs/ERC1155.sol:63]: Same as mintToken

**Review by ngmitam — CHANGES_REQUESTED** [review]

File: contracts/parents/DepositFactoryContract.sol [hunk 19]
```diff
@@ -9,139 +9,437 @@ import "@openzeppelin/contracts/security/Pausable.sol";
...
+                    address(this),
+                    whiteList_
+                );
+                _depositContracts[clone] = true;
+                _depositContractsList.push(clone);
+                deployedDepositContracts[dstChainId][sellerAddress] = clone;
+                emit DepositContractCreated(clone);
+            } else {
+                address masterDepositContractAddress = address(
+                    new DepositContract(
```
Comment (ngmitam) [inline contracts/parents/DepositFactoryContract.sol:175]: Don't use new, it makes the code size increase

**Comment by ngmitam** [comment]
Resolve conflicts

File: contracts/parents/DepositFactoryContract.sol (hunk 19 above)
Comment (NguyetPhamFuix) [reply contracts/parents/DepositFactoryContract.sol:175]: How should we do if we haven't had a master contract to create clone yet?

Comment (ngmitam) [reply contracts/parents/DepositFactoryContract.sol:175]: Deploy master contract then set on the contract

**Comment by ngmitam** [comment]
Please update tasks script and deploy to test nets

**Comment by ngmitam** [comment]
We still need mint function on depositContract, but don't send to layerZero. Please change it.

**Comment by ngmitam** [comment]
Continue other tasks with the master branch. These tasks are completed.

**Pull request closed without merging**